### DISCUSSION TITLE
Vista20P: Receiving data okay. System does not respond to data sent

### DISCUSSION THREAD

**Discussion by emmills602** [Dilbert66/esphome-vistaECP#133] [Q&A]
I very much appreciate all the work that has gone into this Project. Quite incredible. Noob disclaimer.

I am able to monitor the vista system just fine. However, I can send data, either through the keypanel via the IP address or from HA service calls, but there is no response. In essence, commands that are sent are not being implemented. Many thanks!

Used ESP model: 
ESP32. One note, I am using the VistAlarm.yaml DEV version, modified with my particulars, but the ESP32 board, when first flash, creates an esphome yaml that refers to 'esp32dev' - but I am using what is in the DEV yaml, 'nodemcu-32s'

**Used wiring diagram:**
I am using the PCB layout of the ground-isolated version (oshwlab). I used the recommended PINs from the diagram for (22, 21, and 18). Yellow, green, and black wires are terminated to the panel and the PCB. Power to the ESP32 is from the USB, which takes power from a 12v-5V converter. 

**YAML config (the necessary parts): I removed all lines that were commented.** 
```
substitutions:
  name: "vistaalarm" #unique network name, system name
  friendlyName: "Vista Alarm"  #used as the friendly name of your application in HomeAssistant
  panelId: "VistaAlarm"         #used as the service variable name.
  accesscode: !secret accesscode 
  
vista_alarm_panel:
  id: $panelId
  accesscode: $accesscode #Only comes into effect if needed for arming and quickarm is not set
  maxzones: "48" #maximum amount of zones that your panel supports
  maxpartitions: "1" #maximum amount of partitions that your panel supports
  
  rfseriallookup: "0019994:2:66,0818433:4:22,0123456:1:55,0399512:1:17" # serial1:loop#:zone1,serial2:loop#:zone2
 
  defaultpartition: "1" #set to your main partition
  vistadebuglevel: "3" #component debug level for messages

  keypadaddr1: "18" #partition 1 virtual keyapd
  keypadaddr2: "0" #partition 2 virtual keypad. set to 0 to disable
  keypadaddr3: "0" #partition 3 virtual keypad. set to 0 to disable
  
  rxpin: "22" #GPIO pin to use for data receive (yellow line) 
  txpin: "21" #GPIO pin to use for data transmit (green line)
  monitorpin: "18" #GPIO pin to use for monitoring module traffic such as RF or Expanders . Set to -1 to disable
  
  invert_mon: "true"
  invert_rx: "true"
  invert_tx: "true"
  input_mode_rx: INPUT
  input_mode_mon: INPUT
  
  expanderaddr1: "0" # 1st zone expander emulator (4229) address to use . Set to 0 to disable. 
  expanderaddr2: "0" # 2nd expander emulator address to use . Set to 0 to disable. 
  
  relayaddr1: "0" # relay module emulation (4204) addresses. Set to 0 to disable
  relayaddr2: "0"
  relayaddr3: "0"
  relayaddr4: "0"
  
  ttl: "30000"  # time  to live  in ms for zone/fire status before expiring;
  quickarm: "false"
  lrrsupervisor: "false" # set to true if we don't have an LRR monitoring supervisor  we can emulate one to get the statuses
  
  clean_build: "false" #default is false.  only set to true if getting duplication errors in linking step. Once you compile, reset it back to false.  
  
esp32: 
  board: nodemcu-32s
  framework: 
    type: arduino
    version: recommended
 
external_components:
  - source: github://Dilbert66/esphome-components@dev 
    components: [vista_alarm_panel,template_alarm,web_keypad,mg_lib] 
    refresh: 10min

mg_lib:

web_keypad:
  port: 80 
  partitions: 1  # a keypad will be shown for each partition
  log: false
  config_url: https://dilbert66.github.io/config_files/config_vista.yml
  js_url: https://dilbert66.github.io/js_files/www.js
        
esphome:
  name: $name
  friendly_name: $friendlyName
  
  platformio_options:
    build_flags: 
        - "-Wl,-Map,output.map"  
wifi:
  ssid: !secret extension_wifi_ssid
  password: !secret extension_wifi_password

logger:
  baud_rate: 115200
  level: debug

api:
   encryption:
     key: "{{removed}}" 

safe_mode:  
  
ota:
   platform: esphome
   on_begin:
          switch.turn_off: connection_status_switch 

time:
 - platform: sntp

binary_sensor:
  - platform: template_alarm
    id: z1  
    name: "Front door"
    device_class: door

  - platform: template_alarm
    id: z2
    name: "Back door z2"
    device_class: door   

  - platform: template_alarm
    name: "French doors z3"
    id: z3
    device_class: door  
  
  - platform: template_alarm
    id: z4
    name: "Upstairs bathroom window"
    device_class: window   
 
  - platform: template_alarm
    id: z5
    name: "Dining room motion"
    device_class: motion      

  - platform: template_alarm
    id: z6
    name: "Family room motion"
    device_class: motion     

  - platform: template_alarm
    id: z112    
    name: "check relay 12"  
    device_class: problem

  - platform: template_alarm
    id: z108    
    name: "check expander 08"  
    device_class: problem

  - platform: template_alarm
    id: z103    
    name: "check comm device"  
    device_class: problem
     
  - platform: template_alarm
    id: rdy_1
    name: "Ready partiton 1"

  - platform: template_alarm
    id: trbl_1
    name: "Trouble partition 1"
    device_class: problem
 
  - platform: template_alarm
    id: byp_1
    name: "Bypass partition 1"

  - platform: template_alarm
    name: "Away partition 1"
 
  - platform: template_alarm
    id: arm_1  
    name: "Armed partition 1"  
    
  - platform: template_alarm
    id: arms_1  
    name: "Stay partition 1"
  
  - platform: template_alarm
    id: armi_1
    name: "Instant partition 1"
      
  - platform: template_alarm
    id: armn_1  
    name: "Night partition 1" 
  
  - platform: template_alarm
    id: ac  
    name: "AC Status"
    device_class: plug  
  
  - platform: template_alarm
    id: chm_1  
    name: "Chime partition 1"
 
  - platform: template_alarm
    id: alm_1  
    name: "Alarm partition 1"
 
  - platform: template_alarm
    id: bat  
    name: "Battery Status"
    device_class: problem
   
  - platform: template_alarm
    id: fire_1  
    name: "Fire partition 1"  
    device_class: smoke

  - platform: template_alarm
    id: r121  
    name: "Relay1"

  - platform: template_alarm
    id: r122  
    name: "Relay2"   

text_sensor:

  - platform: template_alarm
    id: ss_1  
    name: "System Status partition 1"
    icon: "mdi:shield"

  - platform: template_alarm
    id: lrr  
    name: "Lrr Msg"
    icon: "mdi:alert-box"

  - platform: template_alarm
    id: rf  
    name: "RF Msg"
    icon: "mdi:alert-box"    

  - platform: template_alarm
    name: "Line1 partition 1"
    id: ln1_1    
  - platform: template_alarm
    name: "Line2 partition 1"
    id: ln2_1    

  - platform: template_alarm
    name: "Zone Status" 
    id: zs

  - platform: template_alarm
    name: "Beeps partition 1"  
    id: bp_1

switch:
  - platform: template
    name: "$name Connection"
    id: connection_status_switch
    restore_mode: DISABLED    
    lambda: |-
      return vista.keybusConnected;
    icon: "mdi:shield-link-variant"
    turn_on_action:
      - switch.toggle: restart_switch
    turn_off_action:
      - lambda: |-
         $panelId->stop();
  - platform: restart
    id: restart_switch
  - platform: safe_mode
    name: "Safe Mode"   

debug:
  update_interval: 300s

sensor:
  - platform: debug
    free:
      name: "Heap Free"
    block:
      name: "Heap Max Block"
    loop_time:
      name: "Loop Time"
```
    
**Issue:**
Key presses, either through the webpanel accessed through the IP or from actions in HA, show as being sent to the panel as being written, but there is no response to them. I can send, but nothing happens.

**Installed sensors (if relevant):** 
Listed above.

**What you've tried:**
- Hard to list everything here as I have been at this a few weeks...yes, a few weeks. 
- I have checked wiring a few times. I initially tried this with the ESP8266, but when I didn't get a response from the keypresses, I started over with an ESP32. I suspect that there is something wrong in my esphome yaml file.
- And, I did activate the address 18 at the hardwired key panel. 
- All the sensors are working - doors, window, and motion. 

**Logs:**
Attached one with me sending the installer code. However, it would look the same if I sent the arm code (which I am sending 'xxxx2' to arm).
[logs_esphome-web-7b11c0_logs.txt](https://github.com/user-attachments/files/16920090/logs_esphome-web-7b11c0_logs.txt)

**Comment by Dilbert66** [comment]
I'm not familiar with the pcb you used since that was done by another user so I can't comment on its wiring or setup but the issue you have can only be related to connectivity of  the transmit line circuit or configuration of the keypad address in the panel.  You either have a bad connection/incorrect pin assignment, or you have not configured the keypad address you are using, in the panel configuration as it's stated in the yaml with programs *19x

**Comment by emmills602** [comment]
Thank you for the response. 

**PCB**
I used the PCB referenced in the README.md file for esphome-VistaECP.
> Thanks to [@howardchen3](https://community.home-assistant.io/u/howardchen3) there is a PCB layout of the ground-isolated version available at [oshwlab](https://oshwlab.com/boobeechen/vistaesphome_copy) that could be used to send to a manufacturer. Read more on the site.
I did assume that the PCB layout works, and it is referenced in several other posts. I looked at the diagram on the EasyEDA program and it appears to match the wiring diagram for ground-isolated version.
<img width="710" alt="Screenshot 2024-09-07 at 5 44 06 PM" src="https://github.com/user-attachments/assets/376998ab-4764-4297-aaa8-81bbe3b6767d">

I realize it was designed for an ESP8266, but I am using an ESP32. I soldered on a female Dupont header onto the PCB. I then used Dupont (female-male) cables to connect the ESP32 to the PCB. I had to order 5 PSBs, so I will try another PCB. Before I do, do you see anything wrong with how I am making the connections? 

**Programming the Keypad**
While I did assign a new keypad address, I wonder if I did this wrong. Here is what I did:

1. At the master (hardwired) keypad, I entered the installer code + 800
2. I then entered *191, for address 18
3. I then entered, 10 - I received confirmation beeps
4. I then exited by pressing *99

What isn't clear to me is that I am doing this at the master keypad. I assume that the address for the master is 16. Am I reassigning the master keypad to address 18 by doing the steps above? (I do have another hardwired keypad in another part of the house, which I assume is address 17 - not sure if there is a way to confirm that)

Do you see anything wrong with what I am doing for the programming?

Thanks again.

**Comment by Dilbert66** [comment]
*191 is correct for enabling keypad address 18 as an additional keypad on partition 1.  No this is not reassigning the keypad, just enabling another address  so this should work fine with the existing addresses 16 and 17.  
So the issue then points to a wiring problem or config on the tx circuit.   As to your connections, no I don't see anything wrong with how you're doing it as long as your connections are tight.  If you want to use this pcb you can also use the ESP version of the wemos d1 mini.  It has a compatible footprint so it can fit on the board. Of course, you will need  to change the gpio pin numbers used as they will not match.

**Comment by emmills602** [comment]
I totally appreciate the help. 

I grabbed a new PCB, soldered on wires to the PCB, and used the other end (female Dupont) to connect to the ESP32. Here is how it is wired up:
<img width="273" alt="Screenshot 2024-09-07 at 9 00 48 PM" src="https://github.com/user-attachments/assets/f05ab5ae-3ec9-436f-a545-1ed458ad9e52">

I reverted back to using 21 (rxpin), 22 (txpin), and 18 (monitorpin), and updated the ESP32.

I am still unable to apparently send key presses. 

I'm adding the log file, although prior comments are clear that there is a wiring issue. However, from the log, when 
I sent the command from HA of `4112800` and I see from the log `Writing keys: 4112800 to partition 1`, and the next line is `[CMD:617]: 1969-12-31 16:38  F7 00 00 07 10 08 00 1C 18 02 00 00 44` Actually, that line comes up a lot if I let the log run longer, so the line doesn't seem to be in response to writing the keys. _[side note, it looks like the date is off, if that is what the initial information seems to suggest]_

So, I'm feeling pretty confident about the wiring between the panel and the PCB, the wiring between the PCB and ESP32, and the configuration file is correct (unless it isn't - attached to my original question). 

I wonder, could it be that I damages the ESP32? 

If there is nothing to glean from the connections diagram above, my original configuration file, or from the attached log, I will try another ESP32. If that doesn't work, I will try the WeMos D1 mini as suggested. Does that sound like a sound approach?

Many thanks!

**WeMos D1 Mini**
I found this on Amazon [link](https://www.amazon.com/QCCAN-Internet-ESP8266-Compatible-MicroPython/dp/B0B9XSNZXD/ref=sr_1_18?crid=26TAXEZIX12ZM&dib=eyJ2IjoiMSJ9.73g5waLIh3SLfX4zhzdiPasTjPpldR94VbZME08qBt6gphoTRaR8Z4KsdZM1gJ6m_9TS-CaSA5d4H2lKwKn7qbMDZ2s1rxLQv26nt1K-jwkHlFDvtuH_7FVraMgm-aB5Q3ZkxBxIs_H4xsYCzRNl9o9pDF9eSnSiO4wdTpFM9mto_WqdjFdhlhiyET9zmzCObb7tHjvyVjurfQF8-AjqFzIdBPDmuWi8orj3X5TAACc.mkRfgdYRVPFts22MjqgFhFJ86RHp50zT0fglw9e3CN0&dib_tag=se&keywords=wemos+d1+mini+esp32+nodemcu&qid=1725770573&sprefix=wemos+d1+mini+esp32+nodemcu%2Caps%2C150&sr=8-18) - will this work? Searching for Wemos D1 minim on Amazon comes up with a lot of options and it isn't clear to me which one to buy.

[logs_esphome-web-7b11c0_logs (2).txt](https://github.com/user-attachments/files/16920791/logs_esphome-web-7b11c0_logs.2.txt)

**Comment by Dilbert66** [comment]
You are still not sending any data.  The F7 line is only the normal prompts and flags that are always sent.  It's not a response to your cmds. You will see the contents of the f7 with the following PROMPT: responses.  If you were sending successfully, you would see a bunch of F6 cmds indicating you were talking with the panel.  The "writing keys" log entry indicates that the ESP has received the request to send to the panel and is trying to send but since there are no following F6 cmds, then it was not successful.  

As to the link you found, that is not the right device. That is an esp8266 one. You want one that says it's esp32 or esp-wroom-32 in the description similar to this below:

D1 Mini NodeMCU ESP32 ESP-WROOM-32 WLAN WiFi

**Comment by Dilbert66** [comment]
You wiring points look ok so I'm at a loss as to what the issue is.  Perhaps, another issue could be related to the optocouplers you are using. Some do not have enough gain. I really don't know what else to tell you.  In cases like this, I would use my oscilloscope to ensure that the signals are going through but most people don't have this type of equipment so it's not something I can suggest to you.  I can't imagine that you damaged the esp as they are pretty resilient.  You can always a different pin if you want..   

You also noted, the date being off. That comes from your esphome yaml time config.

**Comment by emmills602** [comment]
I tried a fresh ESP32, with the same results. I did also try returning to 21, in lieu of TX. I just ordered a D1 mini and will give that a shot. 

I feel like I failed to follow the recipe for a PB&J, dozens of times.

I have now tried 2 ESP32s, and one ESP8266, and 3 different PCBs. Given that I am getting the same results, I am wondering if there was a manufacturing error in the making of the PCB. Does this look like there is a resistor R10 missing?
![IMG_8798](https://github.com/user-attachments/assets/97bdf45d-6be5-44ee-b9e9-172f651182ff)

**Comment by Dilbert66** [comment]
Yes it is missing and  I think it's the 180ohm which would explain it! Lol.  All I see are the 4.7kohm ones. Lol, that analogy to a pb&j is hillarious! I hear ya.  I have all the tools to test things instead of doing trial and error so it's easy for me to debug on my own stuff.

**Comment by Dilbert66** [comment]
FYI, you don't need to use exactly 180.  Anything in that ballpark is good up to around 220. Any higher you might start getting signal issues. You don't want to go too low either.

**Comment by emmills602** [comment]
Excellent. 

**Buying the Resistor**
There are options for watts (1/4W, 5W, etc.) and tolerances (I saw 5% and 1%, but I don't know what I need). 

Also, there are metal film fixed resistors, metal oxide fixed film resistors, and I saw one aluminum shells resistors. I am waaay outside my comfort zone here. 

**Installation**
I am assuming that I need to solder each end of the 180ohm resistor to the two contacts on the PCB. Does it matter which end of the resistor? In my picture above, the bottom contact comes from the ESP32/8266 (for transmission) to Optoisolator Transistor (using the manufacturers description here) which eventually connects to the green line, if I am reading the diagram correctly.

Thanks a bunch.

**Comment by Dilbert66** [comment]
You can use any cheap carbon/metal film fixed 1/4 watt resistor with any tolerance (the 1%,5%,20% is the tolerance) since it's a low power application and there is no need for accuracy. Yes, you need to solder each end, doesn't matter what end.  Yes, it connects from the esp gpio21 pin to the  optocoupler.

**Comment by emmills602** [comment]
Thank you so much. I very much appreciate the support. I try not to bother the experts, but after hours and hours of looking for a solution, I just finally had to ask for some help. Thank you for your help and for this amazing project.

**Comment by Dilbert66** [comment]
Post back when you get it working! Glad to be of assistance!

**Reply by emmills602** [reply]
**UPDATE:** 

- Wednesday, as soon as the resistor arrived, I laid down some solder like Jackson Pollock and connected the resistor to the PCB and reinstalled it. But, the logs did not change from before. 
- Today, I took my time and did a nice job soldering a new resistor onto another PCB. I reinstalled it and...drum roll...no change. It's not working. I get the same F7s, a couple FBs, a couple of F2s - no idea what FBs and F2s are but most importantly, **they aren't any F6s.** 

I'm still working under the assumption that I mis-ordered the PCB (somehow), or there was a mistake in the manufacturing (I say that because 1) it isn't working and during production, and 2) one of the boards didn't pass their internal audit - it looked like the corner of one board was broken). But when the lunch tray is dirty on the airplane, you start to wonder if they are doing maintenance on the engines.

- I could try re-ordering the PCB, this time with the 180ohm resistor, but should I trust that the [oshwlab](https://oshwlab.com/boobeechen/vistaesphome_copy) is correct?
- I could beg someone else that has order the PCB, and that might an extra one, to sell it to me - with a minimum order of 5 PCBs, there must be more than one person willing to part ways and to take pity.

I'll take any suggestions, but at this point, I don't want this saga to continue and you have been more than generous with your time. 

Thanks again...I'm going to go make a pb&j.

**Comment by Dilbert66** [comment]
Interesting, we'll I sure would like to see it working for you. From the picture above, I can't see any other issues, perhaps make sure you didnt change the pinout in the yaml from the recommended one just in case also. 
FYI, FB's are messages related to wireless devices and F2's are messages for any AUI keypads you might have. You should not need to reorder pcb's but you might want to re-melt the solder on the pins of the optcouplers  as they seemed a bit sparse in solder.
Also, please take a picture of the bottom and post so we can make sure all the pins are soldered as well as to see the bottom layer.

**Comment by emmills602** [comment]
As I entered into what I hope is a judgement free zone...here is a picture of the resistor I soldered on. I also tried to add some solder on the optcouplers. I am on the opposite side of skilled solderer, but, up until now, all the other projects I soldered have worked. 
![IMG_8819](https://github.com/user-attachments/assets/4d28b5b4-1503-4e15-9e6f-43a2436b07d4)

And here is a photo of the backside. It's not the greatest photo. 
![IMG_8821](https://github.com/user-attachments/assets/19bbb499-e01a-4e7a-9b78-054c4b542c14)

yes, I only soldered the pins that I needed connections to...
(the black marks I added with a sharpie) 

Using GPIO 21 & 22 -confirmed physical connections to ESP32 and in configuration file.

**Comment by Dilbert66** [comment]
ok, I've compared to the schematic board layout on oshwlab and it matches. Your connections look ok.  Can you confirm the value of that resistor? From the color code it looks to be 290 ohm or 280?  That should be ok.

**Reply by emmills602** [reply]
Here is a picture of the labels on the box and the packing the resistors came in.
![IMG_8822](https://github.com/user-attachments/assets/0dcb1134-f845-4779-92ef-8d6f42e7b4b5)

**Comment by Dilbert66** [comment]
Ok, those 5 band resistor colors always mess me up. Ok, it's 180.  Well, honestly I'm at a loss.

**Reply by emmills602** [reply]
Thanks again for all the support and for this entire project. It is really quite amazing. I'm certain I'll eventually get this working. _(I wonder if I could add wires to the PCB - at least the elements of the schematic that relate to transmission of data - in case that portino of the PCB is not connected correctly. I'll noodle on that )_ Super grateful for the help.

**Comment by emmills602** [comment]
I feel obliged to send a quick update. I have everything working now.

I did order new PCBs based on @howardchen3 's design [link](https://community.home-assistant.io/t/ademco-honeywell-vista-esphome-custom-component-with-an-esp32-esp8266/230537/69). However, I was running into the same issue. 

BUT, and here is where my stupidity steps in, I had never previously connected the red wire to the PCB (the ESP32 was powered by USB). I never connected it because I had previously read in [post](https://community.home-assistant.io/t/ademco-honeywell-vista-esphome-custom-component-with-an-esp32-esp8266/230537/145) that connecting the red wire was unnecessary. Turns out, that isn't true or I misinterpreted the comment.

In any event, with the red wire now connect, everything is working well. I remain very grateful to you for developing this project and for your support in troubleshooting. Be well.

**Comment by Dilbert66** [comment]
ahh, yes, that would explain it!  Data is signaled by switching the 12v (red wire) to the green line. That's what the optocoupler does.

**Comment by Dilbert66** [comment]
Thanks for the update on the issue. I was curious what was going on! Appreciate that! Glad it works for you now.
